### PR TITLE
Moves the Execute Preprocessor into a new library!!!

### PR DESCRIPTION
**Pull request by MSeal**
As a part of https://github.com/jupyter/nbconvert/issues/1045 -- more specifically https://github.com/jupyter/nbconvert/issues/821 -- the execute pattern of the execute preprocessor is now importable in it's own package with this PR. This will give a more focused package for executing notebooks that can be updated on a faster release cycle than nbconvert can manage. It'll also provide a lighter weight library for building execution clients ontop of (e.g. papermill).

I believe I've faithfully kept the abilities of the preprocessor, including many of the backwards compatabilities with old nbconvert version behaviors. The tests all copied over with minimal changes to cover this behavior. NBConvert will now import this package and wrap it with the existing preprocessor (likely to change to a "Processor" name in a later stage of the release) rather than implementing notebook execution iteself. 

Next steps:
- Point readthedocs to this repo
- Setup circileci to look at this repo
- Automate a changelog copy to docs page?
- Do the first release (0.1.0?)
- Update nbconvert 6.0 alpha to import this package
- Update papermill 2.0 (master) to import this package

Please comment on the PR for docs, setup info, or obvious code issues. Thanks!

### CODE REVIEW

File: docs/execute.rst [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+    executor = Executor(nb, timeout=600, kernel_name='python3', resources={'metadata': {'path': 'notebooks/'}})
+
+We specified two (optional) arguments ``timeout`` and ``kernel_name``, which
+define respectively the cell execution timeout and the execution kernel.
+Usually you don't need to set these options, but these and other options are
+available to control execution context. Note that ``path`` specifies
+in which folder to execute the notebook.
+
+**Execute/Run (preprocess)**: To actually run the notebook we call the method
+``preprocess``::
```
Comment (choldgraf) [inline docs/execute.rst]:
```suggestion
``execute``::
```

File: docs/execute.rst [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+
+    executor.execute()
+
```
Comment (choldgraf) [inline docs/execute.rst]:
```suggestion

This notebook will now have its cell outputs populated with the result
of running each cell.

```

File: docs/execute.rst [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+Hopefully, we will not get any errors during the notebook execution
+(see the last section for error handling).
```
Comment (choldgraf) [inline docs/execute.rst]:
```suggestion
(see the last section for error handling). This notebook will
now have its cell outputs populated with the result of running
each cell.
```

File: docs/execute.rst [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+there are no execution errors. But, what if there are errors?
+
+Execution until first error
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+An error during the notebook execution, by default, will stop the execution
+and raise a ``CellExecutionError``. Conveniently, the source cell causing
+the error and the original error name and message are also printed.
+After an error, we can still save the notebook as before::
+
+    with open('executed_notebook.ipynb', mode='w', encoding='utf-8') as f:
```
Comment (choldgraf) [inline docs/execute.rst]: Isn't it also possible to just call `write` and a path to a notebook file, and it'll do all the `with open(` stuff under the hood?

**Comment by choldgraf** [comment]
This is awesome! @MSeal thanks for the work to port this over

Since this is a big code contribution, there's 0 files in the repository right now, and this isn't publicly released on PyPI or anything, I'd be +1 on a quick pass for reasonable-ness, and then we merge this so that it's easier to fork / install the package / test out / make iterative PRs and improvements. What do folks think about that?

File: docs/execute.rst (hunk 4 above)
Comment (MSeal) [reply docs/execute.rst]: Yep, I blindly copied this section -- will update.

**Comment by MSeal** [comment]
Given @choldgraf's suggestion and the folks that have taken a look so far, I'm going to merge and make an issue for doing the first release.

**Comment by MSeal** [comment]
Link to new issue: https://github.com/jupyter/nbclient/issues/3

**Pull request merged**